test/integration/data/elastic_torch_main.py

```diff
@@ -0,0 +1,136 @@
+
+import argparse
+import json
+import os
+import time
+
+import torch
+import torch.nn.functional as F
+
+import horovod.torch as hvd
+
+parser = argparse.ArgumentParser(description='PyTorch Elastic Test',
+                                 formatter_class=argparse.ArgumentDefaultsHelpFormatter)
+
+parser.add_argument('--batches-per-epoch', type=int, default=10,
+                    help='number of batches per epoch')
+parser.add_argument('--batches-per-commit', type=int, default=1,
+                    help='number of batches per commit of the elastic state object')
+parser.add_argument('--epochs', type=int, default=3,
+                    help='number of epochs')
+parser.add_argument('--logfile', default='/tmp/logfile.txt',
+                    help='log file to record results (one line per epoch)')
+parser.add_argument('--discovery-schedule', default='[]',
+                    help='JSON string specifying schedule of host updates each epoch')
+parser.add_argument('--exit-schedule',
+                    help='JSON string mapping from (epoch, batch) to list of ranks to exit at that time')
+
+args = parser.parse_args()
+
+hvd.init()
+
+batch_size = 32
+data = torch.randn(batch_size, 2)
+target = torch.LongTensor(batch_size).random_() % 2
+
+lr = 0.001
+model = torch.nn.Sequential(torch.nn.Linear(2, 2))
+optimizer = torch.optim.SGD(model.parameters(), lr=lr * hvd.size())
+optimizer = hvd.DistributedOptimizer(optimizer, named_parameters=model.named_parameters())
+
+hostname = os.environ.get('HOROVOD_HOSTNAME')
+start_rank = int(os.environ.get('HOROVOD_RANK', 0))
+
+discovery_schedule = json.loads(args.discovery_schedule)
+epoch_to_hosts = {epoch: hosts for epoch, hosts in discovery_schedule if epoch is not None}
+default_hosts = discovery_schedule[-1][1] if len(discovery_schedule) > 0 else []
+
+exit_schedule = json.loads(args.exit_schedule) if args.exit_schedule else {}
+
+
+def check_exit(epoch, batch):
+    key = str((epoch, batch))
+    if key in exit_schedule:
+        ranks_to_exit = exit_schedule[key]
+        if start_rank in ranks_to_exit:
+            raise RuntimeError('check_rank and exit epoch={} batch={} start_rank={} rank={}'
+                               .format(epoch, batch, start_rank, hvd.rank()))
+
+
+def log_state(state):
+    state_dict = {
+        'epoch': state.epoch,
+        'batch': state.batch,
+        'commits': state.commits,
+        'hostname': hostname,
+        'start_rank': start_rank,
+        'rank': hvd.rank(),
+        'size': hvd.size(),
+        'rendezvous': state.rendezvous}
+    with open(args.logfile, 'a') as f:
+        f.write(json.dumps(state_dict) + os.linesep)
+
+
+@hvd.elastic.run
+def train(state):
+    state.rendezvous += 1
+    while state.epoch < args.epochs:
+        print('epoch {} batch {}'.format(state.epoch, state.batch))
+
+        while state.batch < args.batches_per_epoch:
+            check_exit(state.epoch, state.batch)
+
+            optimizer.zero_grad()
```

Comment:
should `optimizer` and `model` be accessed through `state`?

Reply:
It doesn't have to be, since we don't re-assign the variable in `state`.